src/compiler/checker.ts

```diff
@@ -16875,8 +16875,8 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                 if (getTypeParameterFromMappedType(mappedType) === getActualTypeVariable(type)) {
                     const typeParameter = getHomomorphicTypeVariable(mappedType);
                     if (typeParameter) {
-                        const constraint = getConstraintOfTypeParameter(typeParameter);
-                        if (constraint && everyType(constraint, isArrayOrTupleType)) {
+                        const constraint = getConstraintOfType(getConditionalFlowTypeOfType(typeParameter, node));
+                        if (constraint && everyType(constraint, t => isArrayOrTupleType(t) || isArrayOrTupleOrIntersection(t))) {

```

Comment:
`isArrayOrTupleOrIntersection` should really be called `isArrayOrTupleIntersection`; I spent too much time trying to figure out if this was redundant.

(You copied this from somewhere, not your doing)

Reply:
(If you were to stick this in the PR I'd be okay with it)

Reply:
ye, that's a good rename - I pushed it out.